File: CS/WindowsFormsApplication6/FilterPanel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Columns;
using DevExpress.Utils.Drawing;
using DevExpress.LookAndFeel;
using DevExpress.Utils;
using DevExpress.Skins;
using DevExpress.XtraEditors.Repository;
using System.Linq;
using DevExpress.XtraGrid.Views.Grid;
using System.Collections;
using DevExpress.XtraLayout;
using DevExpress.Data.Filtering;

namespace WindowsFormsApplication6 {
    public partial class FilterPanel : DevExpress.XtraEditors.XtraUserControl {
        CustomGridView view;
        DevExpress.XtraGrid.Registrator.GridEmbeddedLookAndFeel ElementsLookAndFeel { get { return view.GetType().GetProperty("ElementsLookAndFeel", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance).GetValue(view, null) as DevExpress.XtraGrid.Registrator.GridEmbeddedLookAndFeel; } }
        public bool IsVisible { get { return dictionary != null && dictionary.Count > 0; } }
        public FilterPanel() {
            InitializeComponent();
        }
        Dictionary<GridColumn, BaseEdit> dictionary;
        SimpleButton btn;
        public FilterPanel(CustomGridView view, List<GridColumn> columns)
            : this() {
            this.view = view;
            BackColor = GetBackColor();
            UpdateFilterEditors(columns);
        }
        public void SetColumns(List<GridColumn> columns) {
            if(this.dictionary != null && this.dictionary.Count > 0) this.dictionary.Clear();
            UpdateFilterEditors(columns);
        }
        void DisposeEdits() {
            foreach(var item in dictionary) {

                ((BaseEdit)item.Value).EditValueChanged -= FilterControl_EditValueChanged;
            }
            dictionary.Clear();
        }
        public CriteriaOperatorCollection GetCollection(CriteriaOperator oper) {
            Type t = oper.GetType();
            var prop = t.GetProperty("Operands");
            if(prop != null) {
                return prop.GetValue(oper,null) as CriteriaOperatorCollection;
            }
            return new CriteriaOperatorCollection();
        }
        public void OnFilterChanged(DevExpress.Data.Filtering.CriteriaOperator oper) {
            foreach(var item in dictionary) {
                item.Value.EditValue = null;
            }
            if(oper!=null)Recursion(oper);
        }

        void Recursion(CriteriaOperator oper) {
            if(oper is DevExpress.Data.Filtering.FunctionOperator) {
                GetFunctionalOperands(oper);
            } else {
                foreach(var item in GetCollection(oper)) {
                    Recursion(item);
                }
            }
        }

        void GetFunctionalOperands(CriteriaOperator item) {
            var operands = GetCollection(item);
            
            if(operands.Count == 2) {
                var nameOperand = operands.OfType<OperandProperty>().FirstOrDefault();
                var col = view.Columns.ColumnByFieldName(nameOperand.PropertyName);
                var valOperand = operands.OfType<OperandValue>().FirstOrDefault();
                
                if(col != null) {
                    if(dictionary.ContainsKey(col)) {
                        var editor = dictionary[col];
                        editor.EditValue = valOperand.Value;
                    }
                }
            }
        }


        public void UpdateFilterEditors(List<GridColumn> cols) {
            if(dictionary != null) DisposeEdits();
            ClearLayout();
            var items = new List<DevExpress.XtraLayout.BaseLayoutItem>();
            dictionary = new Dictionary<GridColumn, BaseEdit>();
            int i = 0;
            foreach(var col in cols.OrderBy(c => c.VisibleIndex)) {
                var repositoryEditor = view.GetRepositoryItem(col);
                var editor = repositoryEditor.CreateEditor();
                editor.Text = "";
                if(repositoryEditor is RepositoryItemTextEdit) editor.Size = editor.MaximumSize = editor.MinimumSize = new Size(103, 20);
                else if(repositoryEditor is RepositoryItemCheckEdit) {
                    ((CheckEdit)editor).Properties.NullStyle = DevExpress.XtraEditors.Controls.StyleIndeterminate.Inactive;
                    ((CheckEdit)editor).Properties.AllowGrayed = true;
                    editor.EditValue = null;
                    editor.Size = editor.MaximumSize = editor.MinimumSize = editor.CalcBestSize();
                }

                ((BaseEdit)editor).EditValueChanged += FilterControl_EditValueChanged;
                dictionary.Add(col, (BaseEdit)editor);

                var item = layoutControlGroup1.AddItem();
                editor.Name = Guid.NewGuid().ToString();
                item.Control = editor;
                item.TextAlignMode = DevExpress.XtraLayout.TextAlignModeItem.AutoSize;
                item.Name = col.FieldName;
                if(i != 0) { item.Move(items[i - 1], DevExpress.XtraLayout.Utils.InsertType.Right); }
                items.Add(item);

                i++;
            }
            if(items.Count > 0) {
                var buttonItem = layoutControlGroup1.AddItem();
                if(btn != null) { btn.Dispose(); btn = null; }
                btn = new SimpleButton();
                btn.Name = Guid.NewGuid().ToString();
                btn.Size = btn.MaximumSize = btn.MinimumSize = new Size(60, 22);

                btn.Text = DevExpress.XtraGrid.Localization.GridLocalizer.Active.GetLocalizedString(DevExpress.XtraGrid.Localization.GridStringId.FilterPanelCustomizeButton);
                buttonItem.Control = btn;
                buttonItem.TextVisible = false;
                buttonItem.Move(items.Last(), DevExpress.XtraLayout.Utils.InsertType.Right);
                btn.Click += btn_Click;
                var emptySpaceItem = new EmptySpaceItem();
                emptySpaceItem.Name = "emptySpaceItemNew";
                layoutControl1.Root.AddItem(emptySpaceItem);
                emptySpaceItem.Move(buttonItem, DevExpress.XtraLayout.Utils.InsertType.Right);
                layoutControlGroup1.BestFit();
            }

        }

        void FilterControl_EditValueChanged(object sender, EventArgs e) {
            UpdateFilter(sender);
        }

        void UpdateFilter(object sender) {
            var pair = dictionary.FirstOrDefault(c => c.Value == sender);
            object editingValue = null;
            BaseEdit activeEditor = sender as BaseEdit;
            if(activeEditor != null) {
                editingValue = activeEditor.EditValue;
            } else editingValue = ((Control)sender).Text;
            view.ChangeFilterValue(pair.Key, editingValue);
        }
        void btn_Click(object sender, EventArgs e) {
            view.ShowFilterEditor(view.FocusedColumn);
        }
        public void ClearLayout() {
            layoutControlGroup1.Clear();
        }
        Color GetBackColor() {
            if(ElementsLookAndFeel.ActiveStyle == ActiveLookAndFeelStyle.Skin) {
                SkinElement element = GridSkins.GetSkin(ElementsLookAndFeel)[GridSkins.SkinGridGroupPanel];
                if(element.Color.BackColor2 != Color.Empty)
                    return element.Color.BackColor2;
            }
            return LookAndFeelHelper.GetSystemColor(ElementsLookAndFeel, SystemColors.Control);
        }

    }
}

File: VB/WindowsFormsApplication6/FilterPanel.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.Utils.Drawing
Imports DevExpress.LookAndFeel
Imports DevExpress.Utils
Imports DevExpress.Skins
Imports DevExpress.XtraEditors.Repository
Imports System.Linq
Imports DevExpress.XtraGrid.Views.Grid
Imports System.Collections
Imports DevExpress.XtraLayout
Imports DevExpress.Data.Filtering

Namespace WindowsFormsApplication6
	Partial Public Class FilterPanel
		Inherits DevExpress.XtraEditors.XtraUserControl

		Private view As CustomGridView
		Private ReadOnly Property ElementsLookAndFeel() As DevExpress.XtraGrid.Registrator.GridEmbeddedLookAndFeel
			Get
				Return TryCast(view.GetType().GetProperty("ElementsLookAndFeel", System.Reflection.BindingFlags.NonPublic Or System.Reflection.BindingFlags.Instance).GetValue(view, Nothing), DevExpress.XtraGrid.Registrator.GridEmbeddedLookAndFeel)
			End Get
		End Property
		Public ReadOnly Property IsVisible() As Boolean
			Get
				Return dictionary IsNot Nothing AndAlso dictionary.Count > 0
			End Get
		End Property
		Public Sub New()
			InitializeComponent()
		End Sub
		Private dictionary As Dictionary(Of GridColumn, BaseEdit)
		Private btn As SimpleButton
		Public Sub New(ByVal view As CustomGridView, ByVal columns As List(Of GridColumn))
			Me.New()
			Me.view = view
			BackColor = GetBackColor()
			UpdateFilterEditors(columns)
		End Sub
		Public Sub SetColumns(ByVal columns As List(Of GridColumn))
			If Me.dictionary IsNot Nothing AndAlso Me.dictionary.Count > 0 Then
				Me.dictionary.Clear()
			End If
			UpdateFilterEditors(columns)
		End Sub
		Private Sub DisposeEdits()
			For Each item In dictionary

				RemoveHandler CType(item.Value, BaseEdit).EditValueChanged, AddressOf FilterControl_EditValueChanged
			Next item
			dictionary.Clear()
		End Sub
		Public Function GetCollection(ByVal oper As CriteriaOperator) As CriteriaOperatorCollection
			Dim t As Type = oper.GetType()
			Dim prop = t.GetProperty("Operands")
			If prop IsNot Nothing Then
				Return TryCast(prop.GetValue(oper,Nothing), CriteriaOperatorCollection)
			End If
			Return New CriteriaOperatorCollection()
		End Function
		Public Sub OnFilterChanged(ByVal oper As DevExpress.Data.Filtering.CriteriaOperator)
			For Each item In dictionary
				item.Value.EditValue = Nothing
			Next item
			If oper IsNot Nothing Then
				Recursion(oper)
			End If
		End Sub

		Private Sub Recursion(ByVal oper As CriteriaOperator)
			If TypeOf oper Is DevExpress.Data.Filtering.FunctionOperator Then
				GetFunctionalOperands(oper)
			Else
				For Each item In GetCollection(oper)
					Recursion(item)
				Next item
			End If
		End Sub

		Private Sub GetFunctionalOperands(ByVal item As CriteriaOperator)
			Dim operands = GetCollection(item)

			If operands.Count = 2 Then
				Dim nameOperand = operands.OfType(Of OperandProperty)().FirstOrDefault()
				Dim col = view.Columns.ColumnByFieldName(nameOperand.PropertyName)
				Dim valOperand = operands.OfType(Of OperandValue)().FirstOrDefault()

				If col IsNot Nothing Then
					If dictionary.ContainsKey(col) Then
						Dim editor = dictionary(col)
						editor.EditValue = valOperand.Value
					End If
				End If
			End If
		End Sub


		Public Sub UpdateFilterEditors(ByVal cols As List(Of GridColumn))
			If dictionary IsNot Nothing Then
				DisposeEdits()
			End If
			ClearLayout()
			Dim items = New List(Of DevExpress.XtraLayout.BaseLayoutItem)()
			dictionary = New Dictionary(Of GridColumn, BaseEdit)()
			Dim i As Integer = 0
			For Each col In cols.OrderBy(Function(c) c.VisibleIndex)
				Dim repositoryEditor = view.GetRepositoryItem(col)
				Dim editor = repositoryEditor.CreateEditor()
				editor.Text = ""
				If TypeOf repositoryEditor Is RepositoryItemTextEdit Then
					editor.MinimumSize = New Size(103, 20)
					editor.MaximumSize = editor.MinimumSize
					editor.Size = editor.MaximumSize
				ElseIf TypeOf repositoryEditor Is RepositoryItemCheckEdit Then
					CType(editor, CheckEdit).Properties.NullStyle = DevExpress.XtraEditors.Controls.StyleIndeterminate.Inactive
					CType(editor, CheckEdit).Properties.AllowGrayed = True
					editor.EditValue = Nothing
					editor.MinimumSize = editor.CalcBestSize()
					editor.MaximumSize = editor.MinimumSize
					editor.Size = editor.MaximumSize
				End If

				AddHandler CType(editor, BaseEdit).EditValueChanged, AddressOf FilterControl_EditValueChanged
				dictionary.Add(col, CType(editor, BaseEdit))

				Dim item = layoutControlGroup1.AddItem()
				editor.Name = Guid.NewGuid().ToString()
				item.Control = editor
				item.TextAlignMode = DevExpress.XtraLayout.TextAlignModeItem.AutoSize
				item.Name = col.FieldName
				If i <> 0 Then
					item.Move(items(i - 1), DevExpress.XtraLayout.Utils.InsertType.Right)
				End If
				items.Add(item)

				i += 1
			Next col
			If items.Count > 0 Then
				Dim buttonItem = layoutControlGroup1.AddItem()
				If btn IsNot Nothing Then
					btn.Dispose()
					btn = Nothing
				End If
				btn = New SimpleButton()
				btn.Name = Guid.NewGuid().ToString()
				btn.MinimumSize = New Size(60, 22)
				btn.MaximumSize = btn.MinimumSize
				btn.Size = btn.MaximumSize

				btn.Text = DevExpress.XtraGrid.Localization.GridLocalizer.Active.GetLocalizedString(DevExpress.XtraGrid.Localization.GridStringId.FilterPanelCustomizeButton)
				buttonItem.Control = btn
				buttonItem.TextVisible = False
				buttonItem.Move(items.Last(), DevExpress.XtraLayout.Utils.InsertType.Right)
				AddHandler btn.Click, AddressOf btn_Click
				Dim emptySpaceItem = New EmptySpaceItem()
				emptySpaceItem.Name = "emptySpaceItemNew"
				layoutControl1.Root.AddItem(emptySpaceItem)
				emptySpaceItem.Move(buttonItem, DevExpress.XtraLayout.Utils.InsertType.Right)
				layoutControlGroup1.BestFit()
			End If

		End Sub

		Private Sub FilterControl_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateFilter(sender)
		End Sub

		Private Sub UpdateFilter(ByVal sender As Object)
			Dim pair = dictionary.FirstOrDefault(Function(c) c.Value Is sender)
			Dim editingValue As Object = Nothing
			Dim activeEditor As BaseEdit = TryCast(sender, BaseEdit)
			If activeEditor IsNot Nothing Then
				editingValue = activeEditor.EditValue
			Else
				editingValue = DirectCast(sender, Control).Text
			End If
			view.ChangeFilterValue(pair.Key, editingValue)
		End Sub
		Private Sub btn_Click(ByVal sender As Object, ByVal e As EventArgs)
			view.ShowFilterEditor(view.FocusedColumn)
		End Sub
		Public Sub ClearLayout()
			layoutControlGroup1.Clear()
		End Sub
		Private Function GetBackColor() As Color
			If ElementsLookAndFeel.ActiveStyle = ActiveLookAndFeelStyle.Skin Then
				Dim element As SkinElement = GridSkins.GetSkin(ElementsLookAndFeel)(GridSkins.SkinGridGroupPanel)
				If element.Color.BackColor2 <> Color.Empty Then
					Return element.Color.BackColor2
				End If
			End If
			Return LookAndFeelHelper.GetSystemColor(ElementsLookAndFeel, SystemColors.Control)
		End Function

	End Class
End Namespace
